File: threeSum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  const answer = [];

  const sorted = [...nums].sort();

  let a = 0;
  while (a < sorted.length) {
    if (a > 0 && sorted[a - 1] === sorted[a]) {
      a++;
      continue;
    }
    // find b, c 2 sum problem
    let left = 0;
    let right = sorted.length - 1;

    while (left < right) {
      const threeSum = sorted[a] + sorted[left] + sorted[right];
      if (threeSum > 0) {
        right--;
      } else if (threeSum < 0) {
        left++;
      } else {
        answer.push([sorted[a], sorted[left], sorted[right]]);
        left++;
        while (left < right && sorted[left - 1] === sorted[left]) left++;
      }
    }
    a++;
  }
  return answer;
};
// console.log(threeSum([-1, 0, 1, 2, -1, -4]));
console.log(threeSum([-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4]));

File: JavaWorkSpace/workspace/src/JavaPlayGround/ThreeSum.java
package JavaPlayGround;

import java.util.*;

public class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> output = new LinkedList<List<Integer>>();
        Arrays.sort(nums);

        // Start from a, and look for 2 number b, c
        for (int i = 0; i < nums.length; i++) {
            int a = nums[i];

            if (i > 0 && nums[i - 1] == a)
                continue;

            int left = i + 1, right = nums.length - 1;

            while (left < right) {
                int threeSum = a + nums[left] + nums[right];

                if (threeSum > 0) {
                    right--;
                } else if (threeSum < 0) {
                    left++;
                } else {
                    // 0, pick 3 numbers
                    output.add(List.of(a, nums[left], nums[right]));
                    left++;

                    while (left < right && nums[left] == nums[left - 1])
                        left++;
                }
            }
        }

        return output;
    }
}
